keps/sig-network/3766-referencegrant/README.md

```diff
@@ -0,0 +1,492 @@
+weaken existing namespace boundaries. We believe ReferenceGrant will have a net
+benefit on the ecosystem as it will allow workloads, secrets, and configuration
+to be deployed in separate namespaces that more clearly match up with desired
+authorization.
+
+## Design Details
+
+### General Notes
+
+#### `ReferenceGrant` is half of a handshake
+
+When thinking about ReferenceGrant, it is important to remember that it does not
+do anything by itself. It *Grants* the *possibility* of making a *Reference*
+across namespaces. It's intended that _another object_ (that is, the From object)
+complete the handshake by creating a reference to the referent object (the To
+object).
+
+#### `Kind` vs `Resource`
+
+When creating a metaresource (that is, a resource that targets other resources)
+like ReferenceGrant, it's important to consider if the metaresource uses the more
+common `Kind` or the more correct `Resource`.
+
+When designing the Gateway API in general, we considered this quite a bit (and
+@robscott even ended up making https://github.com/kubernetes/community/pull/5973
+to clarify the API conventions), but in the end decided to use Kind, to improve
+the user experience. That is, it's easier users to take the value from the `kind`
+field at the top of the YAML they are already using, and put it straight into
+these fields, rather than needing to do a kind-resource lookup for every user's
+interaction with the API.
+
+Secondly, as in the Ingress API, the target types must be clearly defined by
+each implementation, so there should be no opportunity for ambiguity.
+
+Lastly, this is only half of a handshake, so the From referent must also both
+support referencing the To referent, and must reference the one in the To field.
+
+With all of that in mind, we felt that the benefits in terms of specificity
+outweighed the costs in terms of user pain to use the more correct `Resource`.
+
+### API Spec
+
+```golang
+// ReferenceGrant identifies kinds of resources in other namespaces that are
+// trusted to reference the specified kinds of resources in the same namespace
+// as the policy.
+//
+// Each ReferenceGrant can be used to represent a unique trust relationship.
+// Additional ReferenceGrants can be used to add to the set of trusted
+// sources of inbound references for the namespace they are defined within.
+//
+// ReferenceGrant is a form of runtime verification allowing users to assert
+// which cross-namespace object references are permitted. Implementations that
+// support ReferenceGrant MUST NOT permit cross-namespace references which have
+// no grant, and MUST respond to the removal of a grant by revoking the access
+// that the grant allowed.
+//
+// Support: Core
+type ReferenceGrant struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of ReferenceGrant.
+	Spec ReferenceGrantSpec `json:"spec,omitempty"`
```

Comment:
Do we think we will really have status?

Reply:
I would like to be able to have a status that indicates which references this resource has enabled and for which implementations. Maybe a simple structure like this:

```yaml
status:
  referencesAllowed:
  - controllerName: gateway.example.com
    numReferences: 1
```

In general we want users to be able to understand if a ReferenceGrant is actually being used or if it can be safely deleted. It would be even better if we could describe the references it enabled, but that would almost certainly not be scalable.

Reply:
This seems like an unbounded list problem.  I guess you could truncate.

You'll need participating controllers to a) have authz for refgrant/staus (probably OK, but a bit scary); and b) periodically refresh their name and count in this list; and c) clean up when the ref is not used any more.  It sounds fragile and racy.

I ACK the desire for visibility (applaud it, actually) but I am not sure this is the right design.

Perhaps what you want is a RefGranted resource which indicates a single reference-grant-use, or maybe even just a periodic event saying "this refgrant is in use by gateway foo in namespace bar".  That's easy to periodically throw, doesn't have an unbounded list, doesn't have RBAC issues or cleanup issues.

I don't know if an accurate tally really matters here, as much as a "someone is using this" ? 

If we're going to bake spec/status into the API, we should believe it is needed, so maybe a bit of legwork on this aspect?

Reply:
> Perhaps what you want is a RefGranted resource which indicates a single reference-grant-use or maybe even just a periodic event saying "this refgrant is in use by gateway foo in namespace bar".

I'm a bit concerned with how noisy events or additional resources would get here. In the case of events, it seems like every controller would need to publish an event for each ReferenceGrant it was using at least 1x an hour. In the case of resources, the number of ReferenceGrantUse resources could become massive. 

If we're using status, I think the unbounded list is slightly less of an issue when the list is just the number of distinct implementations of the API. That seems highly likely to be <100 for a long time, but I could be wrong. We could also take the struct I'm suggesting putting in status and making a resource with just that for ReferenceGrantUse...

>I don't know if an accurate tally really matters here, as much as a "someone is using this" ?

I mostly agree, but a tally could help someone understand the order of magnitude + if they were successfully reducing the number of uses in preparation for a delete.

Reply:
> ```yaml
> status:
>   referencesAllowed:
>   - controllerName: gateway.example.com
>     numReferences: 1
> ```

- If this kind of behavior is important, maybe what a finalizer could do the right job? (If you don't want a controller to be able to set that finalizer, you can choose not to let it).
- I'm thinking that if we want a controller to be able to set `status.referencesAllowed`, we could have a matching, _optional_ declaration within `.spec` about which controller IDs this reference grant applies to.


We don't need `.status` right now, just room to add it later.

Reply:
> I'm a bit concerned with how noisy events or additional resources would get here. In the case of events, it seems like every controller would need to publish an event for each ReferenceGrant it was using at least 1x an hour. In the case of resources, the number of ReferenceGrantUse resources could become massive.

ehhhhh.  We don't have it today, and it took 8 years for us to justify it, so will it REALLY grow that large that quickly?  Events are assumed to be somewhat high N already.  That said, let's define what it is that you're tying to indicate.  Is it "this refgrant was used" or "this refgrant was used approximately N times" or "this refgrant was used EXACTLY N times" or this refgrant is used by Foo X in namespace Y" ?

Those all have different requirements :)

> If we're using status, I think the unbounded list is slightly less of an issue when the list is just the number of distinct implementations of the API

Then a RefGrantUse resource is the same - if it is one per subject per controller, it's the same cardinality, right?

Reply:
I think I have the following goals, in descending order of importance:

1. Communicate that the ReferenceGrant is actively being used
2. Communicate which controllers are using this ReferenceGrant
3. Communicate how many times it's been used with sufficient granularity that I can see the effects of my changes (if I remove this reference, am I removing a dependency on this ReferenceGrant?)

Both 1 and 2 could be accomplished with a finalizer, and each individual implementation could choose to expose finalizers as an option since I realize that many don't like to deal with finalizers. I'm a bit sad to give up on 3, but if we _really_ want something like that we could introduce a secondary resource like you're describing.

Reply:
If we think that the cardinality of controllers is low, we can put it in status.  The downsides are:
1) Could frequently require retries because of optimisitic concurrency failures (I'm trying to increment my count, but so is everyone else)
2) If we're wrong about cardinality, there's not an easy way out
3) Lots of writes to a resource that will be watched fairly often (every controller which needs refs will watch all refgrants)
4) We need `.status`.

If we instead put that into a `ReferenceGrantUse` resource (just a tuple of controller-name and count), then we only have optimisitic concurrency problems with ourselves, we have ~infinite cardinality, nobody will be watching them, and RefGrant doesn't need .status.

Downsides:
1) It's another new resource
2) It's a new pattern, untested in other places.

Reply:
Moved to "open questions" since I think we're going to need more discussion here.